### DISCUSSION TITLE
Introducing wsnic, a fresh take at the websockproxy server

### DISCUSSION THREAD

**Discussion by chschnell** [copy/v86#1199] [Show and tell]
**[wsnic](https://github.com/chschnell/wsnic)** is a WebSocket to TAP device proxy server for linux intended as a fully compatible alternative to [benjamincburns/websockproxy](https://github.com/benjamincburns/websockproxy) (and its derivates). The service provided is identical but it is implemented using different methods.

Benchmark comparison for a 374 MB download from a local HTTP server using curl (websockproxy is unthrottled):

![wsnic-20241222](https://github.com/user-attachments/assets/3aff5426-de79-4538-bdfb-90eb75683f38)

Maximum throughput:

Proxy | Buildroot | Debian 12
:--- | ---: | ---:
websockproxy | 909 kB/s | 1,752 kB/s
wsnic | 32,768 kB/s | 15,360 kB/s

Under Buildroot, wsnic is around 36 times faster than websockproxy, and under Debian around 9 times faster.

Of course, these benchmarks vary widely from system to system, but they do provide reproducible results for each system.

**Comment by spetterman66** [comment]
Well yeah thats great looking forward for when its implemented in v86 :D

**Reply by chschnell** [reply]
> Well yeah thats great looking forward for when its implemented in v86 :D

wsnic uses the same NetworkAdapter in V86 as websockproxy, I run all tests against the current V86 master.

So it works with an unmodified V86, but it is still work in progress so I wouldn't recommend using it yet.

**Comment by copy** [comment]
Thanks, this looks quite promising. Feel free to send a PR to point docs/networking.md to this implementation.

**Reply by chschnell** [reply]
@copy, I saw you noticed my attempt at an overhawl of the networking documentation at [A facelift for the networking documentation #1239](https://github.com/copy/v86/discussions/1239), I could turn it into a PR if you don't mind.

I'm only a bit unsure about introducing the new term *network backend*, I felt the need for it to clearly separate its concept from the other components, but the rest fits well in I believe.

**Reply by copy** [reply]
Yes, please send a PR.

The term network backend sounds good to me (it's also what qemu uses). I am considering going further and replacing `relay_url` with a backend object in the api, as `"inbrowser"` is neither a url nor a relay. e.g.:

```js
new V86({
    net_device: {
        type: "virtio",
        backend_type: "wisp" | "inbrowser" | ...,
        relay_url: "...",
        // OR
        backend: {
            type: "wisp" | "inbrowser" | ...,
            relay_url: "..."
        },
    },
});
```

If you have any thoughts, let me know.

**Reply by chschnell** [reply]
Looks good, I like the first (flat) style a bit better.

Do you consider removing the current URL (with the backend scheme) in favour of separate fields in `net_device`, or to keep it and move its decoding out into the code that calls the V86 constructor in your main web interface?

Maybe some URL decoder helper function (also useful for applications embedding v86) would come in handy here. I like the current URL scheme for being user friendly.

**Comment by chschnell** [comment]
I found a very nice solution for what to use as the [TLS termination proxy](https://en.wikipedia.org/wiki/TLS_termination_proxy) in wsnic.

So far I had used Apache2 for this task which was very efficient but really quite inconvenient when it comes to documenting its setup procedure.

I had tested [`socat`](https://linux.die.net/man/1/socat) to see if it could replace Apache2, it works but is reaching only about a third of the throughput that Apache2 delivers.

However, today I came across **[`stunnel`](https://www.stunnel.org/)**, it fits my use case perfectly and is performing just as good as Apache2! It was really easy to embed and now serves as a built-in method to support secured `wss://` connections in wsnic out of the box, only a certificate is required. For documentation see wsnic's README at https://github.com/chschnell/wsnic. In all, the project has matured quite a bit.

I've noticed that many other proxy servers also rely on an external TLS termination proxy to support `wss://` connections, I think stunnel could be a highly recommended tool for this task.

**Comment by chschnell** [comment]
I've split wsnic internally into a WebSocket frontend and an abstract network backend. I want to try out the many different technologies the Linux kernel has to offer for virtual networking, besides the one I learned from the original websockproxy, hence this approach.

So far I've implemented 4 different backends and I think I got my first alternate network backend to work. That's exciting to me because I hadn't really ever worked with the virtual networking components in Linux before, and this involved quite a bit of learning.

The 4 backends are:

1. **tapdev** -- a single [TAP device](https://docs.kernel.org/networking/tuntap.html) shared by all WebSocket clients with MAC-based packet routing in wsnic
2. **brtap** -- a single [virtual bridge](https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking#bridge) and one exclusive TAP device per WebSocket client
3. **brveth** -- a single virtual bridge and one [veth pair](https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking#veth) per WebSocket client
4. **macvtap** -- one [macvtap device](https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking#macvtap) per WebSocket client

The first backend uses the same virtual network architecture as the original websockproxy. Since it uses only one shared TAP device for all WebSocket connections it must be able to forward each incoming packet from the TAP device to its proper WebSocket connection, it does so based on the MAC addresses in ethernet packets. It also forwards multicast and broadcast packets from the TAP device to all WebSockets, and does a few other things in this context.

The second one is the one I got to work today. Its advantage over tapdev is that it doesn't have to look at each packet for routing purposes, packets in both directions are just blindly passed through. Another advantage is that I have a clear strategy against the backpressure problem that generally arises in proxy servers (when a sender is pushing data faster than it can be forwarded to the receiver causing the data to balloon in wsnic), yet that's still todo. wsnic's job here is reduced to network setup and WebSocket handling (handshake, ping/pong, and the binary frames just contain plain ethernet frames).

The third one almost works and should perform a bit better than the second one. One end of the veth pair is connected to the bridge, and wsnic connects to the other end using a [packet socket](https://man7.org/linux/man-pages/man7/packet.7.html) (AF_PACKET).

The fourth one is a special case, from what I understood it should behave like another NIC of the host (fullly integrated into the host's network, for example automatically using the same DHCP server as the host), but requires support in your external router hardware (so-called "hairpinning"), it's unclear to me how well supported that is. I tested it in two very different physical networks, macvtap doesn't work in my home setup (my router is a FritzBox) but kinda works in the other. So that one is still work-in-progress.

A few other notes:

- A bit to my surprise, the performance in Firefox and Chrome is indistinguishable, both max out at around 35Mb/s.
- Things I test with the backends: vm-to-network communication (ARP, DHCP), vm-to-vm (ping, socat), vm-to-host, vm-to-intranet and vm-to-internet
- DHCP/DNS is provided by [dnsmasq](https://thekelleys.org.uk/dnsmasq/doc.html) or by an wsnic-internal handler (DHCP only).
- I still want to look into traffic shaping in the Linux kernel to see if I can throttle connections this way
- I also plan to release this as a Docker image

So there's still some work left to do, can't say how long it will take.

Any input, specifically about virtual networking in Linux in this context, is very welcome. Useful documentation about this is hard to come by it seems. Like, on Google I found half a dozen or so similar sounding blog articles but with the exact same shell instructions, and then I found the issue on stackoverflow from which all of them seem to have been autogenerated, Yikes!

**Comment by chschnell** [comment]
Good news, everyone!

The first release of wsnic is online, including a Docker container [`chschnell86/wsnic`](https://hub.docker.com/r/chschnell86/wsnic/tags) at Docker Hub.

Please see the **[wsnic Github repository](https://github.com/chschnell/wsnic)** for installation instructions and how to use it.

I removed all experimental code from the project, there's now only a single supported network backend: the bridged TAP devices. I also removed my own DHCP server implementation, `dnsmasq` is better suited for this and now the only option.

I'd be happy about any feedback about this first release, it is my first Docker container and I'd be interested to hear how it works outside my local setup.

**Reply by SuperMaxusa** [reply]
Yay, that's cool! 

> I'd be happy about any feedback about this first release, it is my first Docker container and I'd be interested to hear how it works outside my local setup.

I tried to run on `podman` (Arch), maybe this use case was not expected, but works after few corrections:

1.  When I tried run using a command from README, I got an error while writing `ip_forward`:
```
Traceback (most recent call last):
  File "/opt/wsnic/wsnic/__main__.py", line 312, in main
    server.run()
    ~~~~~~~~~~^^
  File "/opt/wsnic/wsnic/__main__.py", line 153, in run
    sysctl.write('net/ipv4/ip_forward', 1)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/wsnic/wsnic/__init__.py", line 88, in write
    with open(path, 'w') as f_out:
         ~~~~^^^^^^^^^^^
OSError: [Errno 30] Read-only file system: '/proc/sys/net/ipv4/ip_forward'
```

My workaround was to add [`--privileged`](https://docs.podman.io/en/latest/markdown/podman-run.1.html#privileged) flag to the command to run the container.

2. Inside the Podman (or other non-docker container), `wsnic` runs in default mode, not Docker mode. It because Podman doesn't create `/.dockerenv` file and [this check](https://github.com/chschnell/wsnic/blob/f6c788d73c43f8d2ccc7322aef17aee8a7fdd4e9/wsnic/__main__.py#L44) doesn't work, instead Podman creates a `/run/.containerenv` file in a running container: https://github.com/containers/podman/issues/3586#issuecomment-661918679

-------------------------

And regarding the relay itself: 
Windows and Linux work fine, KolibriOS has a problem with the DNS server, I had something similar on `websockproxy` (see https://github.com/copy/v86/issues/815#issuecomment-1469499212), most likely this is a problem related to `dnsmasq`

![kolibrios-wsnic](https://github.com/user-attachments/assets/a6fb9616-1831-4d0d-a50d-6e9911475286)

**Reply by chschnell** [reply]
@SuperMaxusa, thank you very much for your tests, your comments are highly valuable.

Here is what happened when you ran it under `podman`: Since wsnic failed to detect that it was running in a Docker environment (or Docker-like environment, same to me) it also did not skip the `sysctl` calls which were never meant to be called under Docker.

I added a test for `podman` to wsnic.

But I also removed all `sysctl` calls from wsnic, moved this entire matter into the [Troubleshooting](https://github.com/chschnell/wsnic?tab=readme-ov-file#troubleshooting) section of the README where it really belongs because all this can be done externally to wsnic, and I'm actually not quite sure whether it's really needed at all.

This removes the need for `--privileged` (which I prefer), the only special privilege wsnic needs is Docker's `NET_ADMIN` capability.

I've released a new Docker container at Docker Hub with these fixes (and slightly improved log output).

**Reply by chschnell** [reply]
I'm not sure but I think KolibriOS has more issues (I downloaded `kolibri.img` v0.7.7 from [builds.kolibrios.org](https://builds.kolibrios.org/en_US/data/data/) and used v86's `ne2k` adapter).

If you run wsnic with [CLI option](https://github.com/chschnell/wsnic?tab=readme-ov-file#cli-options) `-d 8.8.8.8` it will setup dnsmasq such that it informs DHCP clients to use Google's DNS instead of dnsmasq's built-in DNS. It gave me the same error message, same with `1.1.1.1` (CloudFlare's).

The EAI code `00000002` that KolibriOS reports is presumably [`EAI_AGAIN`](https://git.kolibrios.org/KolibriOS/kolibrios/src/commit/5593d344cd7f50d7c24fad8daa9f7b6577ae5d67/programs/network.inc#L76), which would mean a programming error in KolibriOS unrelated to the network backend (wsnic): `EAI_AGAIN` means that an attempt to make a blocking call on a non-blocking socket was made (which is a non-error which serves only informative purposes to the caller).

But how would such a bug slip through? Does KolibriOS work with *any* v86 network backend?

**Reply by SuperMaxusa** [reply]
Thank you for the new version, works perfectly on Podman!

> The EAI code `00000002` that KolibriOS reports is presumably [`EAI_AGAIN`](https://git.kolibrios.org/KolibriOS/kolibrios/src/commit/5593d344cd7f50d7c24fad8daa9f7b6577ae5d67/programs/network.inc#L76), which would mean a programming error in KolibriOS unrelated to the network backend

Indeed, I got same error on websockproxy (however, on `relay.widgetry.org` it works) and when I tried to do a DNS lookup or ICMP ping, I got this error (on wsnic):
```
IPv4_output: ARP error=FFFFFFFF
IPv4_output: ip=0x0156A8C0
```
`0x0156A8C0` is `192.168.86.1` - default DNS and gateway, guest Linux can ping via ARP/ICMP this IP.
I also checked the captured packages from v86 in WireShark and no DNS packets were sent.

**Reply by chschnell** [reply]
> (however, on relay.widgetry.org it works)

Interesting, I played a bit around with it and it occasionally works.

I had a hunch, so for a test I throttled the bandwidth of wsnic's network bridge `wsbr0` using `tc` (ececute in a second terminal while wsnic is running, [source](https://www.baeldung.com/linux/throttle-bandwidth)):

```bash
sudo tc qdisc add dev wsbr0 root tbf rate 500kbit burst 16kbit latency 50ms
```

And this helps (not in all cases, but I get much better results with this):

![KolibriOS-20250115](https://github.com/user-attachments/assets/a95c37f1-536d-452c-b049-b9e1429f0957)

This is just a hack to see if it makes a difference. If I understand correctly, that `tc` command throttles the outgoing ("egress") traffic from the VM. I'd rather throttle both directions, but that seems a bit more complicated. What this does however is to delay the TCP ACKs comming from the VM (all packets in fact, also DNS requests), and that throttles overall.

I also found issue #815, maybe it has to do with v86's Ne2k implementation like @copy suggested there?

**Reply by SuperMaxusa** [reply]
Hmm, thanks for your tests.

> I had a hunch, so for a test I throttled the bandwidth of wsnic's network bridge wsbr0 using tc 
<...> 
And this helps (not in all cases, but I get much better results with this)

It's strange that throttle bandwidth, looks like the KolibriOS/v86 is dropping packets (or not?)

------------------

I tried to do a small test to confirm this:

Let's say we have 4 (3 emulators + 1 for wsnic services) clients on the network:
 - Gateway: wsnic - `192.168.86.1`
 - Client A: KolibriOS - `192.168.86.2`
 - Client B: Linux - `192.168.86.3`
 - Client C: Windows 98 - `192.168.86.4`
 
If Client A pings any IP (it's doesn't matter, it's a gateway or another VM), we got error:
```
> 192.168.86.x

Pinging to 192.168.86.x with 32 data bytes
Socket error.
```

However, if Client B pings Client A, this works, but with a lost of 10-28% and large latency:
```
~% ping 192.168.86.2
PING 192.168.86.2 (192.168.86.2) 56(84) bytes of data.
64 bytes from 192.168.86.2: icmp_seq=1 ttl=64 time=12039 ms
64 bytes from 192.168.86.2: icmp_seq=2 ttl=64 time=9960 ms
64 bytes from 192.168.86.2: icmp_seq=3 ttl=64 time=7907 ms
64 bytes from 192.168.86.2: icmp_seq=4 ttl=64 time=6043 ms
64 bytes from 192.168.86.2: icmp_seq=5 ttl=64 time=4041 ms
64 bytes from 192.168.86.2: icmp_seq=6 ttl=64 time=4040 ms
64 bytes from 192.168.86.2: icmp_seq=7 ttl=64 time=4034 ms
64 bytes from 192.168.86.2: icmp_seq=8 ttl=64 time=4040 ms
64 bytes from 192.168.86.2: icmp_seq=9 ttl=64 time=4049 ms
64 bytes from 192.168.86.2: icmp_seq=10 ttl=64 time=4088 ms
^C
--- 192.168.86.2 ping statistics ---
14 packets transmitted, 10 received, 28.5714% packet loss, time 21095ms
rtt min/avg/max/mdev = 4034.040/6024.114/12038.840/2804.151 ms, pipe 5
```

Same with Client C -> Client A:
```
Pinging 192.168.86.2 with 32 bytes of data:

Request timed out.
Request timed out.
Request timed out.
Request timed out.

Ping statistics for 192.168.86.2:
    Packets: Sent = 4, Received = 0, Lost = 4 (100% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum =  0ms, Average =  0ms
```

Pinging Client B -> Client C works properly with normal latency:
```
~% ping 192.168.86.4
PING 192.168.86.4 (192.168.86.4) 56(84) bytes of data.
64 bytes from 192.168.86.4: icmp_seq=1 ttl=128 time=189 ms
64 bytes from 192.168.86.4: icmp_seq=2 ttl=128 time=16.1 ms
64 bytes from 192.168.86.4: icmp_seq=3 ttl=128 time=10.5 ms
64 bytes from 192.168.86.4: icmp_seq=4 ttl=128 time=35.4 ms
64 bytes from 192.168.86.4: icmp_seq=5 ttl=128 time=67.7 ms
64 bytes from 192.168.86.4: icmp_seq=6 ttl=128 time=38.2 ms
64 bytes from 192.168.86.4: icmp_seq=7 ttl=128 time=28.9 ms
64 bytes from 192.168.86.4: icmp_seq=8 ttl=128 time=15.1 ms
64 bytes from 192.168.86.4: icmp_seq=9 ttl=128 time=15.8 ms
64 bytes from 192.168.86.4: icmp_seq=10 ttl=128 time=14.1 ms
^C
--- 192.168.86.4 ping statistics ---
10 packets transmitted, 10 received, 0% packet loss, time 9103ms
rtt min/avg/max/mdev = 10.480/43.080/189.060/51.345 ms
```

And pinging gateway works only for Client B and C properly:
```
B -> Gateway:

~% ping 192.168.86.1
PING 192.168.86.1 (192.168.86.1) 56(84) bytes of data.
64 bytes from 192.168.86.1: icmp_seq=1 ttl=64 time=22.6 ms
64 bytes from 192.168.86.1: icmp_seq=2 ttl=64 time=13.0 ms
64 bytes from 192.168.86.1: icmp_seq=3 ttl=64 time=4.92 ms
<...>
64 bytes from 192.168.86.1: icmp_seq=8 ttl=64 time=5.48 ms
64 bytes from 192.168.86.1: icmp_seq=9 ttl=64 time=9.30 ms
64 bytes from 192.168.86.1: icmp_seq=10 ttl=64 time=12.3 ms
^C
--- 192.168.86.1 ping statistics ---
10 packets transmitted, 10 received, 0% packet loss, time 9053ms
rtt min/avg/max/mdev = 4.700/9.016/22.560/5.405 ms

----------------------

C -> Gateway:

Pinging 192.168.86.1 with 32 bytes of data:

Reply from 192.168.86.1: bytes=32 time=10ms TTL=64
Reply from 192.168.86.1: bytes=32 time=17ms TTL=64
Reply from 192.168.86.1: bytes=32 time=8ms TTL=64
Reply from 192.168.86.1: bytes=32 time=7ms TTL=64

Ping statistics for 192.168.86.1:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 7ms, Maximum =  17ms, Average =  10ms
```

Client A throws the same "Socket error".

**Reply by chschnell** [reply]
> It's strange that throttle bandwidth, looks like the KolibriOS/v86 is dropping packets (or not?)

The difference that currently stands out most to me between wsnic and relay.widgetry.org is that the latter is throttled. So I wanted to see what happens when I throttle wsnic, too.

From what I understand, throtteling with `tc` comes down to queing and delaying outgoing ethernet frames, such that the delayed end receives its frames at a maximum frequency that is never exceeded. The network buffers in Linux are deeper than the Mines of Moria, but when they are exhausted then the sending application is signalled to stop (I believe, but I'm really not sure at all here, `tc` is still very new to me and I might be wrong). So all it changes it the timing between frames.

---

> Let's say we have 4 (3 emulators + 1 for wsnic services) clients on the network:
>
> * Gateway: wsnic - 192.168.86.1
> * Client A: KolibriOS - 192.168.86.2
> * Client B: Linux - 192.168.86.3
> * Client C: Windows 98 - 192.168.86.4

[This confused me at first myself: 192.168.86.1 is the IP address I assign to the network bridge that I create in wsnic. The bridge then replies to the pings by itself.]

So pinging B->A works but shows ~30% packet loss, and C->A shows 100% loss.

buildroot shows also ~25% packet loss:

```
~% ping 192.168.86.7
PING 192.168.86.7 (192.168.86.7) 56(84) bytes of data.
64 bytes from 192.168.86.7: icmp_seq=1 ttl=64 time=2028 ms
64 bytes from 192.168.86.7: icmp_seq=2 ttl=64 time=2029 ms
64 bytes from 192.168.86.7: icmp_seq=3 ttl=64 time=2011 ms
64 bytes from 192.168.86.7: icmp_seq=4 ttl=64 time=2007 ms
64 bytes from 192.168.86.7: icmp_seq=5 ttl=64 time=1002 ms
64 bytes from 192.168.86.7: icmp_seq=6 ttl=64 time=1024 ms
^C
--- 192.168.86.7 ping statistics ---
8 packets transmitted, 6 received, 25% packet loss, time 7062ms
rtt min/avg/max/mdev = 1001.854/1683.246/2028.691/474.152 ms, pipe 3
~%
```

Under FreeDOS I don't get any response, even with a 10 second timeout. But mTCP is not the reference here.

So I think we agree: there is something about Kolibri, Ne2k or both.

---

Note that the network subsystem of KolibriOS works flawlessly and is very responsive under qemu, I used (I'm sure this command line can be improved):

```
qemu-system-i386 -m 128 -fda kolibri.img -boot a -vga vmware -net nic,model=rtl8139 -net user
```

I tested `http://copy.sh/` and `http://www.google.com/` (do you happen to know any good public HTTP URLs for tests?).

So that makes me think that maybe, maybe v86's `Ne2k` has more to do with it as it first seemed.

But the case is still open in my books, FreeDOS also uses the Ne2k adapter and it works just fine (DHCP, DNS and TCP).

---

An unrelated sidenote on FreeDOS (I just found that out): Earlier I said that FreeDOS doesn't reply to pings, but I forgot that the mTCP stack is only active when an mTCP application is running.

So if you run something simple like a netcat TCP server in one FreeDOS instance you can ping it from another. The netcat command would for example be:

```
nc -listen 2000
```

Source: [mTCP/USERDOCS/slip.txt](https://github.com/retrohun/mTCP/blob/master/USERDOCS/slip.txt).

**Comment by chschnell** [comment]
To finish this up, here a benchmark comparing 4 different `websockproxy` proxy server implementations (wisp and fetch are not included here):

* [websockproxy](https://github.com/benjamincburns/websockproxy) (used the unthrottled `bellenottelling/websockproxy`)
* [go-websockproxy](https://github.com/gdm85/go-websockproxy)
* [node-relay](https://github.com/krishenriksen/node-relay)
* [wsnic](https://github.com/chschnell/wsnic)

**Description of the 6 benchmark graphs** (with maximum download performance)

| | Left column | Right column |
| --- | :-- | --: |
| **Top row** | inbrowser<br>27,684 kB/s | wsnic<br>36,864 kB/s |
| **Center row** | bellenottelling<br> 934 kB/s | go-websockproxy<br>873 kB/s |
| **Bottom row** | node-relay (1st run)<br>5,323 kB/s | node-relay (5th run)<br>1,493 kB/s |

**Benchmark graphs**

![websockproxy-20250201](https://github.com/user-attachments/assets/75a8983d-9f25-4acf-95c2-29329b98cd53)

**Notes**

* The additional `inbrowser` benchmark is useful to get an idea for the approximate bandwidth that can be expected with v86 in the local setup, a sort of a baseline. It is not the maximum for a single v86 instance running in the browser, but for two instances (HTTP server `socat` in one VM and client `curl` in the other).
* Each benchmark runs the same test, the only difference is payload size: 1024MB for `inbrowser` and `wsnic`, 64MB for all others. All benchmarks use the `buildroot` image. HTTP server is Apache2 in all cases except `inbrowser`, HTTP client is always `curl`.
* All benchmarks were executed sequentially in the same browser session, with a new v86 instance created for each benchmark.
* Two benchmarks for `node-relay` are shown (the 1st run after restarting the proxy server, and the 5th run) because its bandwidth is steadily degrading with each new connection, reproducibly.
* Proxy server `bellenottelling/websockproxy` produces 4-5 times higher bandwidth in the first run after restarting the proxy server, but then runs at the speed shown in every other run, reproducibly.

**Details**

<details>

<summary>1. Bash script "run_websockproxy.sh" used to launch the different proxy servers</summary>

```bash
#!/bin/bash

run_wsnic () {
    docker run --rm --interactive --tty \
        --cap-add=NET_ADMIN \
        --device /dev/net/tun:/dev/net/tun \
        -p 8086:8086 \
        -p 8087:8087 \
        -v ~/cert/cert.crt:/opt/wsnic/cert/cert.crt \
        -v ~/cert/cert.key:/opt/wsnic/cert/cert.key \
        chschnell86/wsnic -i -n "v86.local"
}

run_bellenottelling () {
    # unthrottled version of benjamincburns/jor1k-relay
    #
    docker run --rm --interactive --tty \
        --privileged \
        --device /dev/net/tun:/dev/net/tun \
        -p 8080:80 \
        bellenottelling/websockproxy
}

run_go_websockproxy () {
    # go-websockproxy build instructions:
    #    cd
    #    git clone https://github.com/gdm85/go-websockproxy.git
    #    cd go-websockproxy
    #    git submodule update --init --recursive
    #    go mod init go-websockproxy
    #    go mod tidy
    #    go mod vendor
    #    make
    #
    sudo ~/go-websockproxy/bin/go-websockproxy \
        --listen-address=0.0.0.0:8080 \
        --tap-ipv4=192.168.86.1/24 \
        --log-level=info
}

run_go_websockproxy_dnsmasq () {
    # run this in parallel to run_go_websockproxy() to enable DHCP server "dnsmasq"
    #
    sudo dnsmasq --no-daemon --no-ping --no-hosts \
        --interface=tap0 \
        --bind-interfaces \
        --except-interface=lo \
        --listen-address=192.168.86.1 \
        --dhcp-range=192.168.86.2,192.168.86.254,12h \
        --dhcp-option=option:router,192.168.86.1 \
        --dhcp-option=option:dns-server,192.168.86.1
}

run_node_relay () {
    docker run --rm --interactive --tty \
        --privileged \
        -p 8080:80 \
        krishenriksen/node-relay
}

print_help () {
    echo -e "USE\n"
    echo -e "    $0 BACKEND\n"
    echo -e "BACKEND\n"
    echo -e "    1: wsnic"
    echo -e "    2: bellenottelling"
    echo -e "    3: go-websockproxy"
    echo -e "    4: go-websockproxy-dnsmasq"
    echo -e "    5: node-relay\n"
}

if [[ $# -ne 1 ]]; then
    print_help
    exit 1
fi

case "$1" in
    1)  run_wsnic
        ;;
    2)  run_bellenottelling
        ;;
    3)  run_go_websockproxy
        ;;
    4)  run_go_websockproxy_dnsmasq
        ;;
    5)  run_node_relay
        ;;
    -h) print_help
        ;;
    *)  echo -e "$0: Bad argument \"$1\"\n"
        print_help
        exit 1
        ;;
esac

```

</details>
<details>

<summary>2. PHP script "file.php" used to generate the download payloads</summary>

```php
<?php

function convertToBytes(string $from): ?int {
    $units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB'];
    $number = substr($from, 0, -2);
    $suffix = strtoupper(substr($from,-2));
    if (is_numeric(substr($suffix, 0, 1))) {
        // B or no suffix
        return preg_replace('/[^\d]/', '', $from);
    }
    $exponent = array_flip($units)[$suffix] ?? null;
    if ($exponent === null) {
        return null;
    }
    return $number * (1024 ** $exponent);
}

$size = 1024 * 1024 * 512;
$filename = "file-512MB.bin";

if (isset($_GET["s"])) {
    $size = convertToBytes($_GET["s"]);
    $filename = "file-" . $_GET["s"] . ".bin";
}

$block_size = 1;
$block_count = 1;

if ($size < 1024) {
    $block_size = $size;
}
else if ($size < 1024 * 1024) {
    $block_size = 1024;
    $block_count = intdiv($size, $block_size);
    $size = $block_size * $block_count;
}
else {
    $block_size = 1024 * 1024;
    $block_count = intdiv($size, $block_size);
    $size = $block_size * $block_count;
}

header("Content-Type: application/octet-stream");
header("Content-Disposition: attachment; filename=\"" . $filename . "\"");
header("Content-Length: " . $size);
passthru("dd if=/dev/zero bs=" . $block_size . " count=" . $block_count . " status=none", $err);
exit();

?>
```

</details>

<details>

<summary>3. Console commands to start the HTTP server VM for the inbrowser backend</summary>

```bash
# setup eth0 to 192.168.86.1
ifconfig eth0 up arp 192.168.86.1; ip link set lo up

# spin up DNS server serving this host under the name "build-v86"
echo "build-v86 192.168.86.1" > /etc/dnsd.conf; dnsd -d

# 1024MB HTTP server
socat TCP-L:80,crlf,fork system:'echo -e HTTP/1.1 200 OK;echo -e Content-Length\\\: 1073741824;echo;dd if=/dev/zero bs=1048576 count=1024 status=none'
```

</details>

See this comment at [IP networking under FreeDOS](https://github.com/copy/v86/discussions/1175#discussioncomment-11416486) about the benchmark's source code.
